src/stores/spaces/SpaceStore.ts

```diff
@@ -349,176 +350,335 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
-            const stack = [rootSpace];
-            while (stack.length) {
-                const op = stack.pop();
-                unseen.delete(op);
-                this.getChildSpaces(op.roomId).forEach(space => {
-                    if (unseen.has(space)) {
-                        stack.push(space);
-                    }
-                });
-            }
-        };
+        // Next we need to determine the roots of any remaining full-cycles.
+        // We sort spaces by room ID to force the cycle breaking to be deterministic.
+        const detachedNodes = new Set<Room>(sortBy(joinedSpaces, space => space.roomId));
 
+        // Mark any nodes which are children of our existing root spaces as attached.
         rootSpaces.forEach(rootSpace => {
-            markTreeChildren(rootSpace, detachedNodes);
+            this.markTreeChildren(rootSpace, detachedNodes);
         });
 
         // Handle spaces forming fully cyclical relationships.
-        // In order, assume each detachedNode is a root unless it has already
+        // In order, assume each remaining detachedNode is a root unless it has already
         // been claimed as the child of prior detached node.
         // Work from a copy of the detachedNodes set as it will be mutated as part of this operation.
+        // TODO consider sorting by number of in-refs to favour nodes with fewer parents.
         Array.from(detachedNodes).forEach(detachedNode => {
-            if (!detachedNodes.has(detachedNode)) return;
+            if (!detachedNodes.has(detachedNode)) return; // already claimed, skip
             // declare this detached node a new root, find its children, without ever looping back to it
-            detachedNodes.delete(detachedNode);
-            rootSpaces.push(detachedNode);
-            markTreeChildren(detachedNode, detachedNodes);
-
-            // TODO only consider a detached node a root space if it has no *parents other than the ones forming cycles
+            rootSpaces.push(detachedNode); // consider this node a new root space
+            this.markTreeChildren(detachedNode, detachedNodes); // declare this node and its children attached
         });
 
-        // TODO neither of these handle an A->B->C->A with an additional C->D
-        // detachedNodes.forEach(space => {
-        //     rootSpaces.push(space);
-        // });
+        return rootSpaces;
+    };
+
+    private rebuildSpaceHierarchy = () => {
+        const visibleSpaces = this.matrixClient.getVisibleRooms().filter(r => r.isSpaceRoom());
+        const [joinedSpaces, invitedSpaces] = visibleSpaces.reduce((arr, s) => {
+            switch (s.getMyMembership()) {
+                case "join":
+                    arr[0].push(s);
+                    break;
+                case "invite":
+                    arr[1].push(s);
+                    break;
+            }
+            return arr;
+        }, [[], []] as [Room[], Room[]]);
```

Comment:
This bit takes a few reads to really understand what it's doing, but overall is fine. Suggestion would be to leverage `getEffectiveMembership` even though it doesn't return anything more helpful - just makes it a bit clearer what the overall intent is down the road. Adding names to the intermediate array also helps a bit.

Something like this:

```suggestion
        const [joinedSpaces, invitedSpaces] = visibleSpaces.reduce(([joined, invited], s) => {
            const membership = getEffectiveMembership(s.getMyMembership);
            switch (membership) {
                case EffectiveMembership.Join:
                    joined.push(s);
                    break;
                case EffectiveMembership.Invite:
                    invited.push(s);
                    break;
            }
            return [joined, invited];
        }, [[], []] as [Room[], Room[]]);
```